c_src/erlang_jq_nif.c

```diff
@@ -253,9 +269,71 @@ static ERL_NIF_TERM make_ok_return(ErlNifEnv* env, ERL_NIF_TERM result) {
     return enif_make_tuple2(env, enif_make_atom(env, "ok"), result);
 }
 
+typedef struct {
+    ErlNifMutex* lock;
+    volatile bool consumed_by_process_json_nif;
+    jq_state* state;
+} JQCancelableStateHolderResource;
+
+static ERL_NIF_TERM
+create_jq_resource_nif(
+        ErlNifEnv* env,
+        int argc,
+        const ERL_NIF_TERM argv[]) {
+    module_private_data* module_data = enif_priv_data(env);
+    JQCancelableStateHolderResource* obj =
+        enif_alloc_resource(module_data->jq_state_holder_resource_type,
+                            sizeof(JQCancelableStateHolderResource));
+    obj->state = NULL;
+    static char* lock_name = "jq_nif_resource_lock";
+    obj->lock = enif_mutex_create(lock_name);
+    obj->consumed_by_process_json_nif = false;
+    // Start by locking the lock to make sure jq_state is set before timeout
+    // The lock needs to be unlocked in process_json_with_jq_resource_nif after
+    // the state obj->state field has been set or after an error has been
+    // detected.
+    enif_mutex_lock(obj->lock);
+    ERL_NIF_TERM term = enif_make_resource(env, obj);
+    enif_release_resource(obj);
+    return term;
+}
+
+static ERL_NIF_TERM
+cancel_jq_resource_nif(
+        ErlNifEnv* env,
+        int argc,
+        const ERL_NIF_TERM argv[]) {
+    module_private_data* module_data = enif_priv_data(env);
+    JQCancelableStateHolderResource* obj;
+    if (enif_get_resource(
+                env,
+                argv[0],
+                module_data->jq_state_holder_resource_type,
+                (void**)&obj)) {
+        if (!obj->consumed_by_process_json_nif) {
```

Comment:
In theory yes (see comment above). There should be a memory barrier when we schedule ourselves out so in practice we would eventually see the update but I will fix by using the atomics API mentioned above because I agree that just using a volatile struct field so communicate between threads is not optimal. Thanks!

Reply:
[9ac00a5](https://github.com/emqx/jq/pull/36/commits/9ac00a526593cd5967f73361dc967d573d602e97)